tests/test-request.js

```diff
@@ -0,0 +1,162 @@
+                xhr.onload = function () {
+                  if (xhr.readyState === 4) {
+                    if (xhr.status === 200) {
+                      responseText = xhr.responseText
+                    }
+                    resolve()
+                  }
+                }
+                await xhr.send('ping! "from client"')
+              }
+            })
+          })
+          expect(responseText).to.equal('ping! "from client"')
+        } finally {
+          child.kill()
+        }
+      })
+      it(`should post json data ${asyncReq ? 'asynchronously' : 'synchronously'} (with correct content-length)`, async () => {
+        const child = childProcess.fork(serverScriptPath, serverArgs)
+        try {
+          let responseText = ''
+          await new Promise((resolve) => {
+            child.on('message', async message => {
+              if (message && message.port) {
+                const xhr = new XMLHttpRequest()
+                xhr.open('POST', `${protocol}://localhost:${message.port}/length`, asyncReq, null, null, false)
+                xhr.onload = function () {
+                  if (xhr.readyState === 4) {
+                    if (xhr.status === 200) {
+                      responseText = xhr.responseText
+                    }
+                    resolve()
+                  }
+                }
+                xhr.send(' \n{  "message":  "hello" }  \r\n')
+              }
+            })
+          })
+          expect(responseText).to.equal('30')
+        } finally {
+          child.kill()
+        }
+      })
+    })
+  })
+})
+
+describe('XMLHttpRequest rejectUnauthorized', () => {
+  const serverScriptPath = ospath.join(__dirname, 'server.js')
+  const serverArgs = ['--ssl']
+  ;[true, false].forEach((asyncReq) => {
+    it(`XMLHttpRequest rejectUnauthorized true, ${asyncReq ? 'async' : 'sync'}`, async () => {
+      const child = childProcess.fork(serverScriptPath, serverArgs)
+      try {
+        let responseText = ''
+        await new Promise((resolve, reject) => {
+          child.on('message', async message => {
+            if (message && message.port) {
+              const xhr = new XMLHttpRequest()
+              xhr.open('POST', `https://localhost:${message.port}/echo`, asyncReq)
+              xhr.onload = function () {
+                if (xhr.readyState === 4) {
+                  if (xhr.status === 200) {
+                    responseText = xhr.responseText
+                  }
+                  reject(new Error('request should not succeed'))
+                }
+              }
+              xhr.onerror = function () {
+                responseText = xhr.statusText
+                resolve()
+              }
+              await xhr.send('ping! "from client"')
+            }
+          })
+        })
+        expect(responseText).to.equal('{"code":"CERT_HAS_EXPIRED"}')
+      } finally {
+        child.kill()
```

Comment:
Check that with rejectUnauthorized true (default) we can't communicate over https with a server with an expired certificate.